src/shared/ui/picker.ts

```diff
@@ -145,3 +148,213 @@ export function verifySinglePickerOutput<T extends vscode.QuickPickItem>(choices
+        alwaysShow: true,
+    }
+
+    /**
+     * @param quickPick A `vscode.QuickPick` to grant iterating capabilities
+     * @param populator An IteratingQuickPickPopulator which can call an iterator and return QuickPickItems.
+     */
+    public constructor(
+        private readonly quickPick: vscode.QuickPick<vscode.QuickPickItem>,
+        private readonly populator: IteratingQuickPickPopulator<TResponse>
+    ) {
+        // append buttons specific to iterating quickPick
+        this.quickPick.buttons = [...this.quickPick.buttons, IteratingQuickPickController.REFRESH_BUTTON]
+
+        this.iterator = this.populator.getPickIterator()
+
+        this.activeIterationTime = new Date()
+    }
+
+    /**
+     * Pauses any existing item loading.
+     * Call upon selecting an item in order to pause additional background loading.
+     */
+    public pauseRequests(): void {
+        this.isPaused = true
+    }
+
+    /**
+     * Starts item loading. Restarts from existing state if paused.
+     * Useful for manual pagination, losing quick pick focus (e.g. via external link), throttling, etc.
+     */
+    public startRequests(): void {
+        this.loadItems(false)
+    }
+
+    /**
+     * Resets quick pick's state. Useful for manual refreshes with the same query parameters
+     */
+    public reset(): void {
+        this.loadItems(true)
+    }
+
+    private async loadItems(reset: boolean): Promise<void> {
+        if (reset) {
+            this.activeIterationTime = new Date()
+            this.isDone = false
+            this.populator.reset()
+            this.iterator = this.populator.getPickIterator()
+        }
+        const scopeIterationTime = this.activeIterationTime
+
+        // unpause the loader (if it was paused previously by a selection)
+        this.isPaused = false
+
+        // use a while loop so we have greater control over the iterator.
+        // breaking out of a for..of loop for an iterator will automatically set the iterator to `done`
+        // manual iteration means that we can use the same iterator no matter how many times we call loadItems()
+        while (!this.isDone && !this.isPaused) {
+            this.quickPick.busy = true
+            try {
+                const newItems = await this.iterator.next()
+                if (!newItems) {
+                    this.isDone = true
+                    break
+                }
+                // TODO: Is there a way to append to this ReadOnlyArray so it doesn't constantly pop focus back to the top?
+                // on reset, first iteration should clear the quickPick's items.
+                // should handle cases where call latencies are stable
+                if (scopeIterationTime === this.activeIterationTime) {
+                    this.quickPick.items = reset ? newItems.value : this.quickPick.items.concat(newItems.value)
+                    // nothing else to reset from here on out
+                    reset = false
```

Comment:
It's considered best practice not to reassign function arguments. Also, this line doesn't seem to have an effect

Reply:
It affects the above line (see comment above). The array would be reset every time through the loop if I don't set this to false.

Reply:
Ah, I see. I think if you factor out the reset logic, you can remove the branches and state transitions.

Note that you're effectively transitioning from

`Running` -> `Running+ShouldReset` -> `Running`

You can simplify this by consolidating the reset logic in the reset() method. Then the state is always either `Running`, `Paused` and/or `Done`